File: packages/shared/types/source-category.ts
/**
 * Strongly typed category taxonomy and helpers (TypeScript mirror).
 *
 * This mirrors `data/source_categories.yaml` and the Python model in
 * `packages/shared/models/source_category.py`. It is the canonical taxonomy
 * shared across all projects (Research Knowledge Engine, Personal Knowledge
 * Base, Classroom Artifact Grader, Course Generator, Workflow Automation
 * Platform, Resume Tailoring Engine).
 *
 * The taxonomy is embedded as a typed constant so it can be consumed in
 * browser and server bundles without a YAML parser. When editing the taxonomy,
 * update `data/source_categories.yaml` and keep this file in sync.
 *
 * All category and subcategory names are kebab-case.
 */

export interface SourceCategory {
  /** Human-readable display name. */
  name: string;
  /** One-sentence description of the category's scope. */
  description: string;
  /** Synonyms / lookup aliases used by findCategoryByTag(). */
  tags: string[];
  /** Stable, broad sub-topics (kebab-case). */
  subcategories: string[];
}

export type SourceCategoryTaxonomy = Record<string, SourceCategory>;

export const SOURCE_CATEGORIES = {
  software: {
    name: "Software",
    description: "Software engineering and development topics.",
    tags: ["software", "programming", "engineering", "development", "dev"],
    subcategories: [
      "frontend",
      "backend",
      "cloud",
      "devops",
      "testing",
      "architecture",
      "programming-languages",
      "open-source",
      "web-development",
      "distributed-systems",
    ],
  },
  "computer-science": {
    name: "Computer Science",
    description: "Core computer science concepts and disciplines.",
    tags: ["computer-science", "cs", "engineering"],
    subcategories: [
      "data-structures",
      "algorithms",
      "databases",
      "operating-systems",
      "networking",
      "cybersecurity",
      "software-engineering",
      "artificial-intelligence",
      "theory",
      "compilers",
    ],
  },
  education: {
    name: "Education",
    description: "Higher education and learning systems.",
    tags: ["education", "learning", "edtech"],
    subcategories: [
      "higher-education",
      "instructional-design",
      "accessibility",
      "online-learning",
      "assessment",
      "curriculum-development",
      "student-success",
      "educational-technology",
    ],
  },
  teaching: {
    name: "Teaching",
    description: "Teaching practices and classroom management.",
    tags: ["teaching", "pedagogy", "classroom"],
    subcategories: [
      "assignments",
      "rubrics",
      "grading",
      "active-learning",
      "project-based-learning",
      "discussions",
      "course-design",
      "classroom-management",
    ],
  },
  cybersecurity: {
    name: "Cybersecurity",
    description: "Information security and defensive practices.",
    tags: ["security", "cybersecurity", "infosec"],
    subcategories: [
      "offensive-security",
      "defensive-security",
      "malware",
      "cryptography",
      "incident-response",
      "digital-forensics",
      "standards",
      "compliance",
      "vulnerability-management",
      "threat-intelligence",
    ],
  },
  ai: {
    name: "Artificial Intelligence",
    description: "Artificial intelligence and machine learning.",
    tags: ["ai", "ml", "machine-learning"],
    subcategories: [
      "llms",
      "machine-learning",
      "agents",
      "embeddings",
      "rag",
      "computer-vision",
      "nlp",
      "industry",
      "reinforcement-learning",
      "ai-safety",
    ],
  },
  research: {
    name: "Research",
    description: "Academic and technical research.",
    tags: ["research", "science", "academia"],
    subcategories: [
      "computer-science",
      "engineering",
      "education",
      "methodology",
      "statistics",
      "papers",
      "reproducibility",
      "peer-review",
    ],
  },
  business: {
    name: "Business",
    description: "Organizations and management.",
    tags: ["business", "management", "organization"],
    subcategories: [
      "strategy",
      "leadership",
      "operations",
      "entrepreneurship",
      "product-management",
      "project-management",
      "marketing",
      "sales",
    ],
  },
  finance: {
    name: "Finance",
    description: "Personal and organizational finance.",
    tags: ["finance", "investing", "money"],
    subcategories: [
      "investing",
      "retirement",
      "taxes",
      "real-estate",
      "budgeting",
      "dividends",
      "personal-finance",
      "markets",
    ],
  },
  career: {
    name: "Career",
    description: "Professional development and employment topics.",
    tags: ["career", "jobs", "leadership", "employment"],
    subcategories: [
      "resumes",
      "interviewing",
      "hiring",
      "compensation",
      "leadership",
      "management",
      "job-market",
      "professional-development",
    ],
  },
} as const satisfies SourceCategoryTaxonomy;

/** Union of valid top-level category keys. */
export type CategoryName = keyof typeof SOURCE_CATEGORIES;

// ---------------------------------------------------------------------------
// Helper functions
// ---------------------------------------------------------------------------

/** Return all top-level category keys (kebab-case). */
export function listCategories(): string[] {
  return Object.keys(SOURCE_CATEGORIES);
}

/**
 * Return the subcategories for `category`, or an empty array if the category
 * is unknown.
 */
export function listSubcategories(category: string): string[] {
  const entry = SOURCE_CATEGORIES[category as CategoryName];
  return entry ? [...entry.subcategories] : [];
}

/** Return `true` if `category` is a known top-level category. */
export function validateCategory(category: string): boolean {
  return category in SOURCE_CATEGORIES;
}

/** Return `true` if `subcategory` is valid for `category`. */
export function validateSubcategory(
  category: string,
  subcategory: string,
): boolean {
  const entry = SOURCE_CATEGORIES[category as CategoryName];
  return Boolean(entry) && entry.subcategories.includes(subcategory as never);
}

/**
 * Return the category key whose tags include `tag` (case-insensitive), or
 * `null` if no category lists the tag.
 */
export function findCategoryByTag(tag: string): string | null {
  const needle = tag.trim().toLowerCase();
  for (const [key, entry] of Object.entries(SOURCE_CATEGORIES)) {
    if (
      needle === key.toLowerCase() ||
      entry.tags.some((t) => t.toLowerCase() === needle)
    ) {
      return key;
    }
  }
  return null;
}

File: packages/shared/models/source_category.py
"""Strongly typed category taxonomy model and helpers.

This module loads the canonical taxonomy from ``data/source_categories.yaml`` and
exposes typed access plus validation helpers shared across all projects that
consume the taxonomy (Research Knowledge Engine, Personal Knowledge Base,
Classroom Artifact Grader, Course Generator, Workflow Automation Platform,
Resume Tailoring Engine).

All category and subcategory names are kebab-case.

Public helpers
--------------
list_categories()                      -> list[str]
list_subcategories(category)           -> list[str]
validate_category(category)            -> bool
validate_subcategory(category, sub)    -> bool
find_category_by_tag(tag)              -> str | None
"""

from __future__ import annotations

from functools import lru_cache
from pathlib import Path

import yaml
from pydantic import BaseModel, Field

# Repo root is four levels up: packages/shared/models/source_category.py
_REPO_ROOT = Path(__file__).resolve().parents[3]
DEFAULT_CATEGORIES_PATH = _REPO_ROOT / "data" / "source_categories.yaml"


class SourceCategory(BaseModel):
    """A single top-level category in the taxonomy."""

    name: str
    description: str
    tags: list[str] = Field(default_factory=list)
    subcategories: list[str] = Field(default_factory=list)


class SourceCategoryTaxonomy(BaseModel):
    """The full taxonomy: a mapping of kebab-case key -> SourceCategory."""

    categories: dict[str, SourceCategory] = Field(default_factory=dict)


def load_taxonomy(path: str | Path | None = None) -> SourceCategoryTaxonomy:
    """Load and validate the taxonomy from YAML.

    Args:
        path: Optional override for the taxonomy file location.

    Returns:
        A validated :class:`SourceCategoryTaxonomy`.
    """
    config_path = Path(path) if path is not None else DEFAULT_CATEGORIES_PATH
    with config_path.open("r", encoding="utf-8") as handle:
        data = yaml.safe_load(handle) or {}
    return SourceCategoryTaxonomy.model_validate(data)


@lru_cache(maxsize=1)
def _cached_taxonomy() -> SourceCategoryTaxonomy:
    """Return the taxonomy loaded from the default path, cached per process."""
    return load_taxonomy()


# ---------------------------------------------------------------------------
# Helper functions
# ---------------------------------------------------------------------------


def list_categories() -> list[str]:
    """Return all top-level category keys (kebab-case)."""
    return list(_cached_taxonomy().categories.keys())


def list_subcategories(category: str) -> list[str]:
    """Return the subcategories for ``category``.

    Args:
        category: A top-level category key.

    Returns:
        The list of subcategory names, or an empty list if the category is
        unknown.
    """
    entry = _cached_taxonomy().categories.get(category)
    return list(entry.subcategories) if entry else []


def validate_category(category: str) -> bool:
    """Return ``True`` if ``category`` is a known top-level category."""
    return category in _cached_taxonomy().categories


def validate_subcategory(category: str, subcategory: str) -> bool:
    """Return ``True`` if ``subcategory`` is valid for ``category``."""
    entry = _cached_taxonomy().categories.get(category)
    return bool(entry) and subcategory in entry.subcategories


def find_category_by_tag(tag: str) -> str | None:
    """Return the category key whose tags include ``tag``.

    Matching is case-insensitive. Returns the first matching category key, or
    ``None`` if no category lists the tag.
    """
    needle = tag.strip().lower()
    for key, entry in _cached_taxonomy().categories.items():
        if needle == key.lower() or needle in {t.lower() for t in entry.tags}:
            return key
    return None


__all__ = [
    "SourceCategory",
    "SourceCategoryTaxonomy",
    "DEFAULT_CATEGORIES_PATH",
    "load_taxonomy",
    "list_categories",
    "list_subcategories",
    "validate_category",
    "validate_subcategory",
    "find_category_by_tag",
]
